kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java

```diff
@@ -0,0 +1,489 @@
+    void expireSegment(final SegmentPartition segmentToDelete) {
+      LOG.info("{}[{}] Expiring segment id {}",
+               name, segmentToDelete.tablePartition, segmentToDelete.segmentId);
+
+      final var expiredDocs = segmentWindows.get(segmentToDelete);
+      expiredDocs.drop();
+    }
+  }
+
+  public MongoWindowedTable(
+      final MongoClient client,
+      final String name,
+      final SegmentPartitioner partitioner
+  ) {
+    this.name = name;
+    this.partitioner = partitioner;
+
+    final CodecProvider pojoCodecProvider = PojoCodecProvider.builder().automatic(true).build();
+    final CodecRegistry pojoCodecRegistry = fromRegistries(
+        getDefaultCodecRegistry(),
+        fromProviders(pojoCodecProvider)
+    );
+
+    this.database = client.getDatabase(name).withCodecRegistry(pojoCodecRegistry);
+    this.metadata = database.getCollection(
+        name + METADATA_COLLECTION_SUFFIX,
+        WindowMetadataDoc.class
+    );
+  }
+
+  @Override
+  public String name() {
+    return name;
+  }
+
+  @Override
+  public MongoWindowFlushManager init(
+      final int kafkaPartition
+  ) {
+    final WindowMetadataDoc metaDoc = metadata.findOneAndUpdate(
+        Filters.eq(WindowMetadataDoc.ID, kafkaPartition),
+        Updates.combine(
+            Updates.setOnInsert(WindowMetadataDoc.ID, kafkaPartition),
+            Updates.setOnInsert(WindowMetadataDoc.OFFSET, NO_COMMITTED_OFFSET),
+            Updates.setOnInsert(WindowMetadataDoc.STREAM_TIME, UNINITIALIZED_STREAM_TIME),
+            Updates.inc(WindowMetadataDoc.EPOCH, 1) // will set the value to 1 if it doesn't exist
+        ),
+        new FindOneAndUpdateOptions()
+            .upsert(true)
+            .returnDocument(ReturnDocument.AFTER)
+    );
+
+    if (metaDoc == null) {
+      throw new IllegalStateException("Uninitialized metadata for partition " + kafkaPartition);
+    }
+
+    LOG.info("{}[{}] Retrieved initial metadata {}", name, kafkaPartition, metaDoc);
+
+    final var partitionSegments = new PartitionSegments(metaDoc.streamTime, metaDoc.epoch);
+
+    final var activeSegments = partitioner.activeSegments(kafkaPartition, metaDoc.streamTime);
+    if (activeSegments.isEmpty()) {
+      LOG.info("{}[{}] No active segments for initial streamTime {}",
+               name, kafkaPartition, metaDoc.streamTime);
+    } else {
+      for (final SegmentPartition segmentToCreate : activeSegments) {
+        // Most likely if we have active segments upon initialization, the corresponding
+        // collections and windowStart indices already exist, in which case #getCollection and
+        // #createIndex should be quick and won't rebuild either structure if one already exists
+        // TODO: optimize by using listCollectionNames to skip redundant collection/index creation
+        //  since we should almost always have an index for each collection for all active segments
```

Comment:
🤷 IMO we shouldn't do this unless we determine there's a reason to optimize it - generally my opinion on optimization TODOs is not to add them into the code since we should only do them if experimentally it proves valuable (unless it's also a code or architectural improvement). Not a hill I'll die on though